src/git/http.ts

```diff
@@ -0,0 +1,321 @@
+/**
+ * Reference discovery
+ * Notes:
+ *
+ * Server SHOULD terminate each non-flush line using LF ("\n") terminator;
+ * client MUST NOT complain if there is no terminator.
+ *
+ * The returned response is a pkt-line stream describing each ref and its current value.
+ * The stream MUST be sorted by name according to the C locale ordering.
+ *
+ * If HEAD is a valid ref, HEAD MUST appear as the first advertised ref.
+ * If HEAD is not a valid ref, HEAD MUST NOT appear in the advertisement list at all, but other refs may still appear.
+ *
+ * The stream MUST include capability declarations behind a NUL on the first ref.
+ * The peeled value of a ref (that is "ref^{}") MUST be immediately after the ref itself, if presented.
+ * A conforming server MUST peel the ref if it’s an annotated tag.
+ *
+ *   advertised-refs  =  (no-refs / list-of-refs)
+ *                       *shallow
+ *                       flush-pkt
+ *
+ *   no-refs          =  PKT-LINE(zero-id SP "capabilities^{}"
+ *                       NUL capability-list LF)
+ *
+ *   list-of-refs     =  first-ref *other-ref
+ *   first-ref        =  PKT-LINE(obj-id SP refname
+ *                       NUL capability-list LF)
+ *
+ *   other-ref        =  PKT-LINE(other-tip / other-peeled)
+ *   other-tip        =  obj-id SP refname LF
+ *   other-peeled     =  obj-id SP refname "^{}" LF
+ *
+ *   shallow          =  PKT-LINE("shallow" SP obj-id)
+ *
+ *   capability-list  =  capability *(SP capability)
+ *   capability       =  1*(LC_ALPHA / DIGIT / "-" / "_")
+ *   LC_ALPHA         =  %x61-7A
+ */
+
+import type { CapabilityList, ObjectGenerator } from './types';
+import * as gitUtils from './utils';
+import {ObjectId, ObjectList} from './types';
+import { EncryptedFS } from "encryptedfs";
+import {Buffer} from "buffer";
+import * as utils from '../utils';
+
+/**
+ * Smart ref discovery response looks like
+ *
+ * ```
+ * S: 200 OK
+ * S: Content-Type: application/x-git-upload-pack-advertisement
+ * S: Cache-Control: no-cache
+ * S:
+ * S: 001e# service=git-upload-pack\n
+ * S: 0000
+ * S: 004895dcfa3633004da0049d3d0fa03f80589cbcaf31 refs/heads/maint\0multi_ack\n
+ * S: 003fd049f6c27a2244e12041955e262a404c7faba355 refs/heads/master\n
+ * S: 003c2cb58b79488a98d2721cea644875a8dd0026b115 refs/tags/v1.0\n
+ * S: 003fa3c2e2402b99163d1d59756e5f207ae21cccba4c refs/tags/v1.0^{}\n
+ * S: 0000
```

Comment:
Beware of the parser generator standard.

Reply:
There's examples of that in the `ids`. And there's an old issue describing the whole concept.

Reply:
Still to align with that standard.

That information needs to be put into the development docs.

Polykey-Docs - MatrixAI-Graph.

Reply:
The `claims` domain also uses it.

Reply:
What's the status with this?

Reply:
I modified the parsers to be consistent with how the claims domain does things. I ended up creating an `isX()` and `assertX()` to use.